### DISCUSSION TITLE
Nested Alpine.data with init() and destroy()?

### DISCUSSION THREAD

**Discussion by mildred** [alpinejs/alpine#4133] [1. Help] [answered]
Hello,

I would like to attach two different behaviours on the same DOM node.

Can I do something like this?

```html
<script>
Alpine.data('persist', () => ({
  init() {
    this.querySelector('.counter').textContent = window.localStorage.get("counter") || "?"
  }
}))
Alpine.data('autoHide', () => ({
  init() {
    setTimeout(() => { this.classList.toggle('hidden', true) }, 1000)
  }
}))
</script>

<div x-data="{persist: persist(), autoHide: autoHide()}">
  <span class="counter"></span>
</div>
```

It seems init() function is not called. How could I do such a thing?

edit: the content of the init() functions is just a realistic example. What I'm interested in is a way to use the lifecycle methods init() and destroy() in a more generic way, possibly more than once on a single DOM element. In the past I used Stimulus and its data-controller is great for this single thing.

**Comment by ekwoka** [comment]
No, because neither are putting an `init` at the root.

You can provide an init that calls both.

You might even want just x-bind or Alpine.directivr

**Comment by SimoTod** [comment]
There are a few typos in your example.
- If you use Alpine data in that way, you are likely to mess up with the scope, It should really be used to generate the full data object as per example in the doc
- `this.querySelector` and `this.classList` will error since this is not the dom element, you need `this.$root.querySelector` or `this.$el.querySelector`
- `localStorage.get` is going to error too, the correct syntax is `localStorage.getItem`

Regarding your question, as pointed out above, init needs to be at the root level, so you can't have 2 init functions auto-called in that way by merging 2 objects. 

You can keep them nested and manually add an x-init and call all the function you needs but I suspect what you want to do is to find a way to compose single behaviour and have all your function called automatically if they exist.

This can be done by adding multiple x-data on the single node so each directive will merge is own bit of data and autocall init. Since Alpine runs directly on the real DOM and you can have only one version of the same attribute, you can trick the browser adding some modifier that Alpine will just ignore for x-data. 

You need an empty x-data (Alpine initially looks for note with that attribute so if you just add x-data.1 and x-data.2 Alpine will not pick it up when it starts) then you can add as many x-data as you want with different modifiers, I used numbers but it doesn't really matter, it can be anything you like. Bear in mind that, if you have conflicting names, the last x-data will take precedence.

Note, this is a bit of an hack and not officially documented but it does work. You can find an example here: https://codepen.io/SimoTod/pen/qBwoqqN

**Reply by mildred** [reply]
To avoid each of the data component sharing their private variables with one another, I came with a way to separate them: https://codepen.io/mildred593/pen/QWPmrYQ

```html
<script>
  class Component {
    constructor(name){
      console.log("construct", name)
      this.name = name
    }
    
    alpine_register() {
      let self = this
      return {
        init() {
          console.log("init", self.name)
          self.root_data = this
          self.root_data[self.name] = self._data
          self.init(...arguments)
        },
        destroy() {
          self.destroy(...arguments)
        }
      }
    }
    
    set data(d) {
      this._data = d
      if(this.root_data) this.root_data[this.name] = d
    }
    
    get data(){
      this.root_data[this.name]
    }
  }
  
  function register_component(name, init) {
    Alpine.data(name,  (n, ...args) => new init(n || name, ...args).alpine_register())
  }
  
  class Persist extends Component {
    init() {
      this.counter = window.localStorage.getItem("counter") || "?"
      this.data = {counter: this.counter}
    }
  }
  
  class AutoHide extends Component {
    constructor(){
      super(...arguments)
      this.data = {hidden: false}
    }
    
    init() {
      setTimeout(() => {
        this.data = {hidden: true}
      }, 5000)
    }
  }
  
  document.addEventListener('alpine:init', () => {
    register_component('persist',  Persist)
    register_component('autoHide', AutoHide)
  })
</script>

<div x-data="persist" x-data.2="autoHide" :class="{hidden: autoHide.hidden}">
  counter is <span x-text="persist.counter"></span>
</div>

<script src="https://cdnjs.cloudflare.com/ajax/libs/alpinejs/3.13.8/cdn.js" defer></script>
```

**Comment by mildred** [comment]
Thank you for your answers. You are right, the example code is not a realistic example and there might be some typos.

`x-init` is nice and could be what I needed if there was a way to hook in a destroy function too. And I'd prefer the object to be accessible as any `x-data` object.

I think the best way I could come to do this is to have a compose function that will call all the init and destroy functions.

I did something like this:

```js
  function compose(obj) {
    return {
      ...obj,
      init() {
        for(let key of Object.keys(obj)) {
          obj[key].init()
        }
      },
      destroy() {
        for(let key of Object.keys(obj)) {
          obj[key].destroy()
        }
      }
    }
  }
```

The only issue I get is that reactivity does not work within the composed objects...

In pen https://codepen.io/mildred593/pen/NWmYYZr I'm using this compose function but I should find a way to extend Alpine's reactivity magic to child objects if possible.

edit: it would be great if attributes modifiers (whish I didn't knew existed) could be used to scope the data:

```html
<div x-data.persist="persist" x-data.autoHide="autoHide" :class="{hidden: $autoHide.hidden}">
  <span x-text="$persist.counter"></span>
</div>
```

**Reply by ekwoka** [reply]
The answer was already provided:

use multiple `x-data` attributes with arbitrary modifiers.

or your own directives.

**Reply by mildred** [reply]
I'd prefer if the different objects would not mix their properties causing possible undefined behaviour...

**Reply by mildred** [reply]
I'm getting a better compose function like this which can preserve reactivity:

```js
    Alpine.data('compose', (obj) => {
      let res = {
        init() {
          for(let key of Object.keys(obj)) {
            this[key].init()
          }
        },
        destroy() {
          for(let key of Object.keys(obj)) {
            this[key].destroy()
          }
        }
      }
      for(let k of Object.keys(obj)) {
        res[k] = new Proxy(obj[k], {
          set(obj, prop, value, receiver) {
            res[k] = receiver
            return Reflect.set(...arguments);
          }
        })
      }
      return res
    })
```

https://codepen.io/mildred593/pen/gOyezGr

**Reply by SimoTod** [reply]
You probably don't need to make it that complex. just user the original object which is already reactive and merge the inits/destoys if you want to go down that route
```javascript
Alpine.data('compose', (obj) => {
  obj.init = function() {
    for(let key of Object.keys(obj)) {
      this[key].init()
    }
  }
  obj.destroy = function() {
    for(let key of Object.keys(obj)) {
      this[key].destroy()
    }
  }
  return obj
})
```

**Reply by mildred** [reply]
I don't think this would work...

- calling init() and destroy() for each of `Object.keys(obj)` works only if there is only composed data objects. Plain value objects will fail
- how would updating keys in sub-objects would trigger reactivity and update the view?

**Reply by SimoTod** [reply, marked as answer]
It's the same as your original example. for a production ready function you'll need additional check. 
The output of alpine data is already reactive by default as long as you don't mess up with the scope internally (your first attempt was creating a different object but calling init from a sub property of the original object). In general, the more you try to hack the internals by adding more proxies, copying data, etc, the more likely you will break something.

The example I posted works, as far as I can see: https://codepen.io/SimoTod/pen/WNWzyGK

**Reply by mildred** [reply]
Ok, I understand, `obj` is not to be discarded because this is the object that is reactive. Thank you.

Putting here your solution (in case the codepen stops working at a future date):

```html
<script>
  document.addEventListener('alpine:init', () => {
    Alpine.data('compose', (obj) => {
      const isInitiable = (param) => {
        return typeof param === 'object'
        && param !== null
        && param.hasOwnProperty('init')
      }

      const isDestroyable = (param) => {
        return typeof param === 'object'
        && param !== null
        && param.hasOwnProperty('destory')
      }

      obj.init = function() {
        for(let key of Object.keys(obj)) {
          if (isInitiable(this[key])) this[key].init()
        }
      }
      obj.destroy = function() {
        for(let key of Object.keys(obj)) {
          if (isDestroyable(this[key])) this[key].destroy()
        }
      }
      obj.debug = function(){
        return JSON.stringify(obj)
      }

      return obj
    })

    Alpine.data('persist', () => ({
      init() {
        this.counter = window.localStorage.getItem("counter") || "?"
        console.log("counter fetched")
      },
    }))
    Alpine.data('autoHide', () => ({
      init() {
        console.log("hiding in progress")
        this.shown = true
        setTimeout(() => {
          this.shown = false
          console.log("hidden")
        }, 5000)
      }
    })) 
  })
</script>

<div x-data="compose({persist: persist(), autoHide: autoHide(), foo: 'bar', baz: {foo: 'bar'}})" :class="{hidden: !autoHide.shown}">
  <span x-text="persist.counter"></span> |
  <span x-text="autoHide.shown"></span> |
  <span x-text="debug"></span>
  <button @click="autoHide.shown = true">Show again</button>
</div>

<script src="https://cdnjs.cloudflare.com/ajax/libs/alpinejs/3.13.8/cdn.js" defer></script>
```

**Reply by ekwoka** [reply]
Just use `this`, not `obj`
